lib/grppc.gi

```diff
@@ -11,6 +11,8 @@
 ##  This file contains the methods for groups with a polycyclic collector.
 ##
 
+# gaplint: disable = all
```

Comment:
I'd be happy to discuss relaxing or turning off certain rules if they are onerous for you, e.g. because you are using an editor that can not easily be configured to avoid them automatically -- so unlike e.g. emacs, vim, ... which all can be configured to honor our our `.editorconfig`, see <https://editorconfig.org>)

But the linting rule which prevents use of tabs is essential to me. We just spent a lot of work resolving the mess of files which used tabs of width 2, 3, 4 or 8, sometimes multiple tab widths mixed in one file or in some cases even in one function, making it really annoying to edit (or even *read*) those functions, and I'd rather make sure no tabs are re-introduced. Surely everyone contributing can configure their editor to not use tabs, at least when editing GAP code?

CC @james-d-mitchell @ChrisJefferson

Reply:
I agree completely @fingolfin

Reply:
Maybe to elaborate a little more, I completely get that having to jump through extra hoops to make contributions to GAP is inconvenient for you @hulpke, and I'm sorry about that. 

I hope it will be possible to convince you that the (hopefully small!) additional cost to you will be a net saving for the GAP project overall. The point of these rules is to make it more convenient for people other than the author of the changes to read and maintain their contributions. There are currently only 4 rule enabled, and, as @fingolfin mentioned, it is more or less trivial to adhere to the 3 out of the 4.

1. No trailing whitespace
2. No tabs
3. No unused local variables
4. No variables assigned but not used.

As @fingolfin mentioned, the first three are more or less trivial to get your editor to do automatically (in vim for example, it's possible to strip trailing whitespace when a file is written, for the tab key to insert a given number of spaces, and there's a vim script function that detects local variables and puts them in the file:
https://github.com/gap-system/gap/blob/master/etc/vim/gap.vim)

No trailing whitespace is for consistency and to avoid masking small changes in PRs that have lots of whitespace changes; no tabs is to make it easier for the same bit of code to be viewed the same way in different editors; the reason for no unused local variables is hopefully obvious (when making the changes in #5275, I found it particularly perplexing that there were some functions with 10s of unused local variables, some of which had comments about what they were for).

The last rule is harder to detect, but as can be seen in #5275, there were lots of instances of unused code in the GAP library, which were detected by the "no variables assigned but not used" rule. The assignment of non-required variables makes the code harder to read and understand, and are often artefacts of refactoring or initial attempts to write the code.

It seems to be fairly common place for large (and not so large) software projects to have and enforce coding conventions of the type mentioned above.  I think there are good reasons for this in those projects and here, as I outlined above, and hoped that the requirements mentioned above are rather minimal, and would improve the quality and maintainability of the GAP code base. 

I'm happy to discuss this further if necessary.

Reply:
Dear @fingolfin 

I completely understand the issue of tabs and now added lines to my .vimrc to handle trailing blanks.

My issue is with requiring validation by a tool in which more and more features get turned on one-by-one. This looks like a solution in search of a problem, or at best like futzing around.

So, if we can limit the requirement to tabs and trailing blanks, no objection.

Once it comes to further checks (maybe just planned and not yet turned on) I think this is overstepping what is meaningful and is actively discouraging contribution.

Reply:
I have no intention to have further checks enabled, and I agree that this quickly turns into futzing around. For me, it's really about the tabs (BTW I have no principle problem with tabs either, just with mixing them with spaces), I am mostly neutral about the trailing whitespaces, and I certainly don't want to "normalize" whitespaces around operators and whatnot

As to the warnings about missing locals: I find them useful, and relatively easy to fix. But if they bother you, I am OK with turning them off again; from my POV they (or rather, @james-d-mitchell -- many thanks!) did their main job for now, i.e., allowing us to eliminate a bunch of dead code, misleading comments on unused variables, and pointless computations. We could also just run them once a year or whatever to see if any meaningful ones have cropped up.